File: Chapter3/Control.cs
namespace SampleCSharp
{
    public abstract class Control
    {
        private int top;
        private int left;

        protected int Top
        {
            get { return this.top; }
            set { this.top = value; }
        }
        protected int Left
        {
            get { return this.left; }
            set { this.left = value; }
        }

        public Control(int top, int left)
        {
            Top = top;
            Left = left;
        }

        //abstract method harus diimplementasikan
        public abstract void DrawWindow();
    }


    public abstract class Window : Control
    {
        public int Right { get; set; }
        public int Bottom { get; set; }

        public Window(int top, int left, int right, int bottom) : base(top, left)
        {
            Right = right;
            Bottom = bottom;
        }

        public abstract void DrawControl();
    }


    public class ListBox : Control, IStorable, ICompressable
    {
        private string listBoxContent;
        public string ListBoxContent
        {
            get { return listBoxContent; }
            set { listBoxContent = value; }
        }

        private int status;
        public int Status
        {
            get { return status; }
            set
            {
                status = value;
            }
        }


        public ListBox(int top, int left, string listBoxContent) : base(top, left)
        {
            this.listBoxContent = listBoxContent;
        }

        public override void DrawWindow()
        {
            Console.WriteLine("Position :({0},{1})", Top, Left);
            Console.WriteLine("Menuliskan string pada listbox: {0}",
            listBoxContent);
        }

        public void Read()
        {
            Console.WriteLine("Reading ListBox...");
        }

        public void Write()
        {
            Console.WriteLine("Writing ListBox...");
        }

        public void Compress()
        {
            throw new NotImplementedException();
        }

        public void Decompress()
        {
            throw new NotImplementedException();
        }
    }

    public class Button : Control
    {
        private string fontColor;
        public Button(int top, int left) : base(top, left)
        {
        }

        public string FontColor
        {
            get { return fontColor; }
            set { fontColor = value; }
        }

        public override void DrawWindow()
        {
            Console.WriteLine("Position :({0},{1})", Top, Left);
            Console.WriteLine("Mewarnai tombol dengan warna: {0}", fontColor);
        }
    }

    public class GridData : Window
    {
        public GridData(int top, int left, int right, int bottom) : base(top, left, right, bottom)
        {

        }

        public override void DrawControl()
        {
            throw new NotImplementedException();
        }

        public override void DrawWindow()
        {
            throw new NotImplementedException();
        }

        public void DrawWindow(string hello)
        {
            Console.WriteLine($"{hello}");
        }
    }
}

File: Chapter3_VB/Control.vb


Public MustInherit Class Control
    Private _top As Integer
    Public Property Top() As Integer
        Get
            Return _top
        End Get
        Set(ByVal value As Integer)
            _top = value
        End Set
    End Property

    Private _left As String
    Public Property Left() As String
        Get
            Return _left
        End Get
        Set(ByVal value As String)
            _left = value
        End Set
    End Property

    Public Sub New(_top As Integer, _left As String)
        Me._top = _top
        Me._left = _left
    End Sub

    Public MustOverride Sub DrawWindow()
End Class

Public Class ListBox
    Inherits Control

    Private _listBoxContent As String
    Public Property ListBoxContent() As String
        Get
            Return _listBoxContent
        End Get
        Set(ByVal value As String)
            _listBoxContent = value
        End Set
    End Property

    Public Sub New(_top As Integer, _left As String, _listBoxContent As String)
        MyBase.New(_top, _left)
        Me._listBoxContent = _listBoxContent
    End Sub

    Public Overrides Sub DrawWindow()
        Console.WriteLine("Drawing ListBox at " & Me.Top & " and " & Me.Left & " with content " & Me.ListBoxContent)
    End Sub
End Class

Public Class Button
    Inherits Control

    Private _buttonContent As String
    Public Property ButtonContent() As String
        Get
            Return _buttonContent
        End Get
        Set(ByVal value As String)
            _buttonContent = value
        End Set
    End Property

    Public Sub New(_top As Integer, _left As String, _buttonContent As String)
        MyBase.New(_top, _left)
        Me._buttonContent = _buttonContent
    End Sub

    Public Overrides Sub DrawWindow()
        Console.WriteLine("Drawing Button at " & Me.Top & " and " & Me.Left & " with content " & Me.ButtonContent)
    End Sub
End Class
